### PR TITLE
feat(spice): recognize and seed the activation boundary

### PR DESCRIPTION
**Pull request by viviveevee**
Extracted from #16411

A chain that upgrades into spice mid-life has a "last pre-spice block". This teaches the node to recognize that block and to record what spice needs about it.

- `spice::boundary` gains the predicates the rest of the boundary work keys on: `is_last_pre_spice_block`, `last_pre_spice_block_header`, and the `spice_producers_epoch_id` / `spice_tracking_prev_hash` routing helpers that point at the first spice epoch for boundary data.
- Block postprocessing seeds the spice execution heads when a first spice block lands, and an uncertified-chunks row for the last pre-spice block so its chunks can certify. Seeding is idempotent across sibling forks.
- The core reader reads that seeded row back and stops ancestry walks for core statements at the boundary, since nothing below it carries any.
- The spice message gate stops dropping messages about the last pre-spice block, and buffers instead of dropping when activation is imminent at the head.

### CODE REVIEW

File: chain/chain/src/spice/core_writer_actor.rs [hunk 1]
```diff
@@ -587,6 +593,21 @@ impl SpiceCoreWriterActor {
         self.send_execution_result_endorsements(&block);
         Ok(())
     }
+
+    fn record_pending_endorsements_for_last_pre_spice_block(
```
Comment (viviveevee) [inline chain/chain/src/spice/core_writer_actor.rs]: move to boundary?

File: chain/chain/src/spice/activation.rs [hunk 2]
```diff
@@ -114,22 +142,24 @@ impl SpiceMessageGate {
...
         kind: SpiceMessageKind,
         block_hash: &CryptoHash,
         unit: DropUnit,
     ) -> bool {
-        let enabled = match spice_enabled_for_block(chain_store, block_hash) {
+        let enabled = match spice_relevant_block(chain_store, epoch_manager, block_hash) {
             Ok(enabled) => enabled,
-            Err(_) => match spice_enabled_at_head(chain_store) {
+            Err(_) => match spice_enabled_or_imminent_at_head(chain_store, epoch_manager) {
                 Ok(enabled) => enabled,
```
[the remark this answers is not in the document]
Comment (viviveevee) [reply chain/chain/src/spice/activation.rs:163]: pre-existing todo(spice)

File: chain/client/src/spice/chunk_validator_actor.rs [hunk 3]
```diff
@@ -734,14 +742,19 @@ impl SpiceChunkValidatorActor {
...
+        // The producers holding the code are those that executed the chunk, which for a
+        // boundary chunk of the last pre-spice block are the first spice epoch's.
+        let producers_epoch_id =
+            spice_producers_epoch_id(self.epoch_manager.as_ref(), &chunk_id.block_hash)?;
         // TODO(spice-data-distribution): this always asks the first producer; spread requesters
         // over the producers and retry with another one if it does not answer.
         let sender = self
             .epoch_manager
-            .get_epoch_chunk_producers_for_shard(&epoch_id, chunk_id.shard_id)?
+            .get_epoch_chunk_producers_for_shard(&producers_epoch_id, chunk_id.shard_id)?
```
[the remark this answers is not in the document]
Comment (viviveevee) [reply chain/client/src/spice/chunk_validator_actor.rs:753]: false positive

File: chain/chain/src/spice/boundary.rs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    }
+    Ok(header)
+}
+
+/// Seeds what the activation boundary needs when `block` is a last pre-spice block
+/// or a first spice block; a no-op otherwise.
+///
+/// A first spice block seeds its parent's uncertified-chunks row again. The parent is
+/// a last pre-spice block by definition, and its own postprocessing may have run on a
+/// binary without this seeding: a node that upgrades late stalls at the first spice
```
Comment (darioush) [inline chain/chain/src/spice/boundary.rs:62]: Can a node reach this case? `check_protocol_version` (`chain_update.rs:83`) panics before commit when a later epoch's version is above the binary's `PROTOCOL_VERSION`, so a binary without spice never commits the last pre-spice block.

The re-seed also does not reach `record_uncertified_chunks_for_block` for this block: the row is only merged here, and that function reads the committed store (`core.rs:1022`). `test_seeding_at_the_first_spice_block_writes_the_missing_parent_row` commits between the two, so it passes. It fails if it follows the order in `chain_update.rs`: seed into the update, merge it, call `record_uncertified_chunks_for_block` on the same update, commit once, then check that the first spice block's row contains the parent's chunks.

Could we drop the re-seed, and instead return an error there when a spice block's pre-spice parent has no row?

```rust
let prev_header = chain_store_update.get_previous_header(block.header())?;
if !prev_header.is_genesis() && !prev_header.is_spice() && uncertified_chunks.is_empty() {
    return Err(Error::Other(format!(
        "missing uncertified chunks of last pre-spice block {prev_hash}"
    )));
}
```

🤖 _Comment prepared using AI._

File: chain/chain/src/spice/core.rs [hunk 5]
```diff
@@ -957,8 +977,10 @@ fn get_uncertified_chunks(
 ) -> Result<Vec<SpiceUncertifiedChunkInfo>, Error> {
     let block = chain_store.get_block(block_hash)?;
 
-    if block.header().is_genesis() || !block.is_spice_block() {
+    if block.header().is_genesis() {
         Ok(vec![])
+    } else if !block.is_spice_block() {
+        Ok(seeded_uncertified_chunks(chain_store, block_hash))
```
Comment (darioush) [inline chain/chain/src/spice/core.rs]: The boundary row is fully determined by the epoch manager and the header. Could this branch compute it instead of persisting it?

```rust
} else if !block.is_spice_block() {
    if !is_last_pre_spice_block(epoch_manager, block_hash)? {
        return Ok(vec![]);
    }
    boundary_uncertified_chunks(epoch_manager, block.header())
}
```

That would remove the write ordering at the first spice block and the case where a missing row reads as certified. It would also replace the `debug_assert` in `seeded_uncertified_chunks`, which a spice block row with no carried-over chunks also passes.

🤖 _-- Claude on behalf of Darioush_

File: chain/chain/src/spice/boundary.rs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        DBCol::uncertified_chunks(),
+        header.hash().as_ref(),
+        &uncertified_chunks,
+    );
+    Ok(())
+}
+
+/// The seeded uncertified-chunks row of the pre-spice `block_hash`: present only for
+/// a last pre-spice block, empty otherwise.
+pub(crate) fn seeded_uncertified_chunks(
```
Comment (darioush) [inline chain/chain/src/spice/boundary.rs:149]: nit: I suggest a name like `get_uncertified_chunks_of_pre_spice_block` or `get_boundary_uncertified_chunks_or_empty`, since the production callsite calls it for all pre-spice blocks.

File: chain/chain/src/spice/boundary.rs [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    }
+}
+
+/// The prev hash shard tracking of `block`'s spice applications is keyed on: `block`
+/// itself for a last pre-spice block, whose chunks are bootstrapped by the shards
+/// tracked in the first spice epoch.
+///
+/// TODO(spice-resharding): for a last pre-spice block the shard tracker resolves shard
+/// ids of the previous layout against the first spice epoch; see the module docs.
+pub fn spice_tracking_prev_hash(
```
Comment (darioush) [inline chain/chain/src/spice/boundary.rs:190]: Why should we special case the last pre-spice block here and in `spice_producers_epoch_id`? If L is the last pre-spice block, the chunk producers of L's own epoch surely applied L and have its `ChunkExtra`. Then it seems this function is not necessary?

If spice activates at epoch S, `should_apply_chunk(IsCaughtUp, L, ..)` is true for shards tracked in S or S+1. So a node that tracks the shard only in S+1 is assumed to have applied L, but it did not.

File: chain/chain/src/spice/core.rs [hunk 8]
```diff
@@ -342,6 +348,20 @@ impl SpiceCoreReader {
         Ok(Some(merklize(&state_roots).0))
     }
 
+    /// Where an ancestry walk for core statements from `block_hash` stops: the last
+    /// certified block, or the last pre-spice block when the certified one is still
+    /// pre-spice, since no core statements exist at or below the activation boundary.
+    fn get_last_certified_block_header_or_last_pre_spice_block(
```
Comment (darioush) [inline chain/chain/src/spice/core.rs:354]: Is this helper needed? When `last_certified` is pre-spice it is the parent of the last pre-spice block, and the old walk to it only adds the last pre-spice block, whose `spice_core_statements()` is empty. Both walks seem to give the same result.

🤖 _Comment prepared using AI._

File: chain/chain/src/spice/activation.rs [hunk 9]
```diff
@@ -42,10 +45,40 @@ pub fn spice_enabled_for_block(
...
+    chain_store: &ChainStoreAdapter,
+    epoch_manager: &dyn EpochManagerAdapter,
+) -> Result<bool, Error> {
+    let head = chain_store.head()?;
+    let next_epoch_protocol_version =
+        epoch_manager.get_next_epoch_protocol_version(&head.last_block_hash)?;
+    Ok(ProtocolFeature::Spice.enabled(next_epoch_protocol_version))
+}
+
+fn spice_enabled_or_imminent_at_head(
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs:62]: nit: optionally consider names like `spice_enabled_at_head_or_next_epoch` and `is_spice_or_last_pre_spice_block`

File: chain/chain/src/spice/activation.rs [hunk 10]
```diff
@@ -177,13 +207,231 @@ impl SpiceMessageGate {
...
+    use std::sync::Arc;
+    use strum::IntoEnumIterator;
+
+    const EPOCH_LENGTH: u64 = 2;
+    const NUM_BLOCKS: usize = 12;
+
+    /// A chain whose headers are all on disk as pre-spice headers, paired with an
+    /// epoch manager that recorded the same blocks with every block after genesis
+    /// voting `vote`.
+    fn setup_gated_chain(vote: ProtocolVersion) -> (Chain, EpochManagerHandle, Vec<Arc<Block>>) {
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs:239]: Could these tests use `setup_pre_spice_chain_with_epoch_length` and `grow_to_last_pre_spice_block` from `spice::tests`? Then the gate reads the chain's own epoch manager, and the second `EpochManager` and `last_pre_spice_index` go away. 

🤖 _Comment prepared using AI._

File: chain/chain/src/spice/tests/mod.rs [hunk 11]
```diff
@@ -2,3 +2,173 @@ mod all_stake_fallback;
...
+        .epoch_manager
+        .add_validator_proposals(block_info, *block.header().random_value())
+        .unwrap();
+    store_update.merge(epoch_manager_update.into());
+    store_update.commit().unwrap();
+}
+
+/// A chain with a pre-spice genesis, for the activation-boundary tests. Its blocks
+/// are fabricated with [`add_pre_spice_block`], not processed.
+pub(crate) fn setup_pre_spice_chain(num_shards: NumShards) -> Chain {
```
Comment (darioush) [inline chain/chain/src/spice/tests/mod.rs:52]: Please consider to move the helpers to pre_spice.rs and also stack the PRs (or merge other PR to master then resolve conflicts on this PR).

File: chain/client/src/spice/tests/chunk_executor_actor.rs [hunk 12]
```diff
@@ -1422,3 +1429,167 @@ fn test_is_descendant_of_final_execution_head_returns_false_for_final_execution_
...
+
+/// Re-seeding the execution heads from sibling boundary forks changes nothing:
+/// both setters are forward-only, and every first spice block on a same-height
+/// fork resolves the same last final block.
+#[test]
+#[cfg_attr(not(feature = "protocol_feature_spice"), ignore)]
+fn test_activation_head_seeding_is_idempotent_across_sibling_boundary_forks() {
+    let (mut chain, signer, blocks) = pre_spice_boundary_chain();
+    let spice_protocol_version = ProtocolFeature::Spice.protocol_version();
+    let last_pre_spice = blocks[3].clone();
```
Comment (darioush) [inline chain/client/src/spice/tests/chunk_executor_actor.rs:1491]: nit: can we avoid hardcoding 3 either by passing a named param to pre_spice_boundary_chain() or using the last block?

File: chain/client/src/spice/tests/chunk_executor_actor.rs [hunk 13]
```diff
@@ -1422,3 +1429,167 @@ fn test_is_descendant_of_final_execution_head_returns_false_for_final_execution_
...
+        build_saved_block(&mut chain, last_pre_spice.as_ref(), 4, spice_protocol_version, &signer);
+    seed_execution_heads(&chain, last_pre_spice.as_ref(), first_spice.as_ref());
+
+    // The canonical first spice block descends through the seeded head's height.
+    assert!(is_descendant_of_final_execution_head(&chain.chain_store, first_spice.header()));
+
+    // A same-height sibling boundary fork passes either way.
+    let sibling_last_pre_spice = build_saved_block(
+        &mut chain,
+        blocks[2].clone().as_ref(),
```
Comment (darioush) [inline chain/client/src/spice/tests/chunk_executor_actor.rs:1563]: nit: In my understanding, "the head" in the doc comment's "walking a block's ancestry down to the head's height" refers to the final execution head, `blocks[2]`.

The sibling fork is built on `blocks[2]`. Would building it on a new block at height 2 (a child of `blocks[1]`) test the case the doc describes?

File: chain/client/src/spice/data_distributor_actor.rs [hunk 14]
```diff
@@ -638,17 +639,18 @@ impl SpiceDataDistributorActor {
         data_id: &SpiceDataIdentifier,
         block: &Block,
     ) -> Result<(HashSet<AccountId>, Vec<AccountId>), Error> {
+        let producers_epoch_id =
+            spice_producers_epoch_id(self.epoch_manager.as_ref(), block.hash())?;
```
Comment (darioush) [inline chain/client/src/spice/data_distributor_actor.rs:643]: For L, this now expects producers from the first spice epoch, but `receive_data_with_block` (line 732) still looks up the sender's key in L's own epoch, and `possible_epoch_ids` (line 933) returns only L's epoch once L is on disk. A producer of the first spice epoch that is not a validator in L's epoch, or whose key changed, would have its boundary data rejected. Should both use `spice_producers_epoch_id`?

(This is another argument to avoid the epoch shifting.)

🤖 _Comment prepared using AI._

**Review by darioush — COMMENTED** [review]
My main question is https://github.com/near/nearcore/pull/16456#discussion_r4088095787, and it seems to be the source of several corner cases.

Happy to discuss it offline and get feedback from @ssavenko-near as well.

I added some smaller comments. To keep e2e verification, I suggest rebasing the full PR #16411 on top of this and adding test-loops to #16411 where the validator set and validator keys change at spice activation.
